Sources/Main.cpp

```diff
@@ -587,6 +587,7 @@ void __fastcall TX584Form::CodeListViewMouseDown(TObject *Sender,
     TListItem *Item = CodeListView->Items->Item[Row];
     //проверяем, не зажата ли клавиша Shift
     if (Shift.Contains(ssShift)) {
+        SelStart = CodeListView->ItemFocused->Index;
```

Comment:
I'm not super excited about adding yet another property to the already bloated `TX584Form`. This often increases complexity and introduces more bugs. You would need to carefully check all selection-related methods and properly initialize `SelStart` everywhere, otherwise you would just add more bugs. For instance, selection can be also changed via Shift+Up and Shift+Down but I don't think the PR handles this case.

The methods `CopyItemClick` and `PasteItemClick` always use `CodeListView->ItemFocused->Index` to determine the current selection. They seem to work pretty well. Probably we can replicate their behavior to fix annoying crashes. What about the following helper method that returns the current selection range from `SelStart` to `SelEnd` **non-inclusive**?

```cpp
void TX584Form::GetSelection(int &SelStart, int &SelEnd)
{
    if (SelCount > 0) {
        SelStart = CodeListView->ItemFocused->Index;
        SelEnd = SelStart + SelCount;
    } else {
        SelStart = CodeListView->ItemFocused->Index + SelCount;
        SelEnd = SelStart + (1 - SelCount);
    }
}
```

You can put this method to section "РЕДАКТОР КОДА" and use it to fix crashes in `CodeTreeViewDblClick` and `DeleteItemClick`. Would this work?

Reply:
I think that I should test this. Thank you for notice

Reply:
I have deleted `SelStart` field and integrated your method `GetSelection` inside the codebase